lib/datadog/security/contrib/rack/patcher.rb

```diff
@@ -0,0 +1,99 @@
+require 'datadog/security/contrib/patcher'
+require 'datadog/security/contrib/rack/integration'
+
+module Datadog
+  module Security
+    module Contrib
+      module Rack
+        module Patcher
+          include Datadog::Security::Contrib::Patcher
+
+          module_function
+
+          def patched?
+            Patcher.instance_variable_get(:@patched)
+          end
+
+          def target_version
+            Integration.version
+          end
+
+          # TODO: logger
+          def logger
+            return @logger if @logger
+
+            @logger ||= ::Logger.new(STDOUT)
+            #@logger.level = ::Logger::DEBUG
+            @logger.level = ::Logger::DEBUG
+            @logger.debug { 'logger enabled' }
+            @logger
+          end
+
+          def patch
+            Instrumentation.gateway.watch('rack.request') do |request|
+              block = false
+
+              Reactive::Operation.new('rack.request') do |op|
+                if defined?(Datadog::Tracer) && Datadog.respond_to?(:tracer) && (tracer = Datadog.tracer)
+                  root_span = Datadog.tracer.active_root_span
+                  active_span = Datadog.tracer.active_span
+
+                  logger.debug { "root span: #{root_span.span_id}"} if root_span
+                  logger.debug { "active span: #{active_span.span_id}"} if active_span
+                end
```

Comment:
It seems like you could abstract the tracer-specific logic out of the `Reactive::Operation`.

These highlighted lines, alongside these
```ruby
                if block && active_span
                  active_span.set_tag('security.blocked', true)
                  active_span.set_tag(Datadog::Ext::ManualTracing::TAG_KEEP, true)
                end
```
may be able to live outside of this context, maybe as an augmentation to `Reactive::Operation` (e.g. `Reactive::TracedOperation`?) with tracer-specific code possibly wrapping the operation code.

Reply:
This sounds sensible. I did not have enough hindsight to do that yet, and preferred to focus on the general AppSec structure part.